cli/commons/src/Microsoft.Kiota.Cli.Commons/IO/ODataPagingService.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+
+[assembly: InternalsVisibleTo("Microsoft.Kiota.Cli.Commons.Tests")]
+namespace Microsoft.Kiota.Cli.Commons.IO;
+
+/// <summary>
+/// Paging service that supports the x-ms-pageable extension
+/// </summary>
+public sealed class ODataPagingService : BasePagingService
+{
+    /// <inheritdoc />
+    public override async Task<Uri?> GetNextPageLinkAsync(PageLinkData pageLinkData, CancellationToken cancellationToken = default)
+    {
+        if (IsJson(pageLinkData))
+        {
+            try
+            {
+                var doc = await JsonDocument.ParseAsync(pageLinkData.Response, cancellationToken: cancellationToken);
+                var hasNextLink = doc.RootElement.TryGetProperty(pageLinkData.NextLinkName, out var nextLink);
+                if (hasNextLink && nextLink.ValueKind == JsonValueKind.String)
+                {
+                    string? link = nextLink.GetString();
+                    if (!string.IsNullOrWhiteSpace(link)) return new Uri(link);
+                }
+            }
+            catch (JsonException)
+            {
+                // If the response isn't valid JSON, there will be no next link.
+                // TODO: Log warning once logging story is defined
+            }
+        }
+
+        return null;
+    }
+
+    /// <inheritdoc />
+    public override async Task<Stream?> MergePageAsync(Stream? currentResult, PageLinkData newPageData, CancellationToken cancellationToken = default)
+    {
+        if (IsJson(newPageData))
+        {
+            return await MergeJsonStreamsAsync(currentResult, newPageData.Response, newPageData.ItemName, newPageData.NextLinkName, cancellationToken);
+        }
+
+        return null;
+    }
+
+    /// <inheritdoc />
+    public override bool OnBeforeGetPagedData(PageLinkData pageLinkData, bool fetchAllPages = false)
+    {
+        // Set the page size to 999 if the user asked to fetch all pages and top either isn't specified or is invalid
+        if (fetchAllPages && (!pageLinkData.RequestInformation.QueryParameters.TryGetValue("%24top", out var topVal) || (topVal as int?) < 1))
+        {
+            pageLinkData.RequestInformation.QueryParameters["%24top"] = 999;
+        }
+        return true;
+    }
+
+    private bool IsJson(PageLinkData pageLinkData)
+    {
+        return pageLinkData.RequestInformation.Headers.TryGetValue("Accept", out var accepts) && accepts.Contains("json");
```

Comment:
Do we need a corresponding change in the abstractions lib to ensure the `Accept` header is set? Otherwise, the check could lead to some unintended consequences.

Reply:
There's an open issue for this.